### DISCUSSION TITLE
Appropriate parameters for homogeneous reconstruction

### DISCUSSION THREAD

**Discussion by olibclarke** [zhonge/cryodrgn#159] [Show and tell]
Hi,

I'm just getting started with CryoDRGN, so apologies if this is basic or if I am doing something stupid. To start with I figured I would try cryoDRGN ab initio homogeneous reconstruction, on a dataset I am familiar with (https://www.ebi.ac.uk/empiar/EMPIAR-11043/).

I first downsampled the particles from the consensus reconstruction like so:

`cryodrgn downsample cryosparc_P1_J3493_013_particles.cs -D 128 -o ank_downsample_128.mrcs`

Then, parsed CTF:

`cryodrgn parse_ctf_csparc cryosparc_P1_J3493_013_particles.cs -o ank_ctf.pkl`

Created a random subset of 20k images:

`cryodrgn_utils select_random 710437 -n 20000 -o ank_downsample_20k.pkl`

Then, ran ab initio reconstruction:

`cryodrgn abinit_homo ank_downsample_128.mrcs --ind ank_downsample_20k.pkl --ctf ank_ctf.pkl -o ./abinit_homo_test >> abinit_homo_test.log`

Resulting reconstruction:
https://www.dropbox.com/s/cfi6fusr5rvc9qq/cryodrgn_reconstruct.mrc?dl=0

As a comparison, I ran single class ab initio reconstruction in cryoSPARC, using default parameters, on a 20k random subset, also downsampled to 128px:

https://www.dropbox.com/s/bv3no0mbwe4uh1s/cryosparc_reconstruct.mrc?dl=0

For comparison, here is a direct reconstruction from the 20k subset, using the poses from the consensus refinement:

https://www.dropbox.com/s/zj6txx9zaa9j6f0/cryosparc_reconstruct_consensus_poses.mrc?dl=0

The results from cryosparc look a lot more detailed, and a lot closer to the consensus refinement. This makes me think I am doing something wrong. Is there something I should alter, in terms of either preprocessing or parameters for the ab initio reconstruction? Many thanks for any help or advice!

<img width="1141" alt="cryodrgn_side" src="https://user-images.githubusercontent.com/19766818/197575736-8e0f45bb-e0ff-40d0-a88d-d3f6730125fc.png">
<img width="1138" alt="cryodrgn_top" src="https://user-images.githubusercontent.com/19766818/197575523-dbd5c99d-398c-4022-bde7-50abac78cd9e.png">
<img width="1143" alt="cryosparc_side" src="https://user-images.githubusercontent.com/19766818/197575549-ffb3dec7-2f61-4f41-b318-286d6beb5506.png">
<img width="1145" alt="cryosparc_top" src="https://user-images.githubusercontent.com/19766818/197575567-32572e02-2913-423e-bb74-b198a8d062ea.png">

**Comment by zhonge** [comment]
Thanks for doing this comparison and sharing the results! I was just thinking about testing your EMPIAR entry myself.

Your commands look right -- I suspect that the default parameters to `cryodrgn abinit_homo` need to be updated, or maybe you need to train for longer? Could you try continuing the training run with `--load latest -n 60`?  I'm also thinking that 20k images might be too few. How long did it take to train by the way?

**Reply by olibclarke** [reply]
Hi Ellen, thanks so much for the quick reply! It took about 40min for training using the default parameters (30 epochs). I will try continuing the training, although comparing the intermediate reconstructions it doesn't seem to be changing much in the later epochs. I wonder if maybe the preprocessing parameters need alteration? These are quite tightly cropped particles, so I'm wondering if some signal is getting masked out?

**Reply by zhonge** [reply]
By default, the pose search only happens every 5 epochs (--ps-freq 5), and with this box size and number of images, I don't think the volume would change much over the last five epochs (25-29). You could also try rerunning with modified pose search settings (maybe like `--ps-freq 2 --l-ramp-epochs 10`), though extending to `-n 60` is probably the easiest test to see if more pose search updates are needed.

**Reply by olibclarke** [reply]
Got it. Trying now with this command: `cryodrgn abinit_homo ank_downsample_128.mrcs --ind ank_downsample_20k.pkl --load latest -n 60 --ctf ank_ctf.pkl -o ./abinit_homo_test >> abinit_homo_test_continue.log`

**Reply by olibclarke** [reply]
(by the way would love to see how it goes in your hands! It is quite a different and I think complementary case to those tried in the cryodrgn papers - it is heterogeneous in both composition and conformation, with flexible elements, and as it is a membrane protein in detergent there is also a different kind of variability introduced by the detergent micelle)

**Reply by olibclarke** [reply]
So 60 iterations definitely gives a better result! There do seem to be some "streaking" artefacts in the micelle which persist, and in general the micelle density seems a lot stronger than we see with ab initio in cryosparc (the big blob in the corner of the box is a secondary micelle, which we do see in 2D classes but only at very low threshold after cryosparc ab initio).
https://www.dropbox.com/s/znzm93ku4pmkju3/cryodrgn_reconstruct_it60.mrc?dl=0

<img width="1085" alt="Screen Shot 2022-10-24 at 2 54 58 PM" src="https://user-images.githubusercontent.com/19766818/197604281-b5a51b49-2a5b-4c7a-b633-25295792e861.png">
<img width="1084" alt="Screen Shot 2022-10-24 at 2 54 30 PM" src="https://user-images.githubusercontent.com/19766818/197604307-857f0aa7-2c70-40fc-ba10-9a2f3abd96b8.png">
<img width="1085" alt="Screen Shot 2022-10-24 at 2 54 17 PM" src="https://user-images.githubusercontent.com/19766818/197604324-038e9ea0-0e55-4b66-98ab-271dcd0f861a.png">

**Reply by olibclarke** [reply]
Looking at individual slices though, there are clear "streaky/spiky" artefacts, as well as an artefactual density extending through the center of the box:
<img width="1055" alt="image" src="https://user-images.githubusercontent.com/19766818/197621619-66197bf9-8d18-44d4-849b-1cc84b7d43a5.png">

**Reply by zhonge** [reply]
Very cool -- glad to see that extending the training helped! IIRC there is an adaptive masking procedure in cryoSPARC -- it could be that cryoDRGN (no masking) produces the average structure including both micelles instead of focusing on the primary one.

I'm not sure what is going on with the streakiness, but we can take a look into it. Thanks for reporting!

**Reply by olibclarke** [reply]
Thanks Ellen - there is adaptive masking in cryosparc, but to my knowledge that is only during refinement, not during ab initio reconstruction, so I don't think that's the reason. Thanks again!

**Reply by olibclarke** [reply]
Ok - so after running a round of `train_vae` and eliminating a small population of outliers (Z>10, 600 particles out of 20k), the results of `abinit_homo` with 60 iterations are much better:
<img width="750" alt="image" src="https://user-images.githubusercontent.com/19766818/197664496-eee59200-80c5-4d4d-93f7-d94616777f0b.png">

**Comment by zhonge** [comment]
The box size should be parsed automatically by `cryodrgn parse_ctf_csparc`, so I'm not sure that's the problem

**Reply by olibclarke** [reply]
Yep - just re-ran with CTF params parsed as suggested in the docs, and doesn't seem to have made any difference
